### DISCUSSION TITLE
ORJSONResponse can't encode subclasses of encodable types

### DISCUSSION THREAD

**Discussion by Yolley** [fastapi/fastapi#9161] [Questions]
### Describe the bug

By design orjson.dumps can't encode subclasses of encodable types, so for example `AnyUrl` from pydantic, which is a subclass of str, can not be encoded with ORJSONResponse
`orjson.JSONEncodeError: Type is not JSON serializable: AnyUrl`

This is an intended behaviour in orjson package, because it is not a drop-in replacement for json (https://github.com/ijl/orjson/issues/1, https://github.com/ijl/orjson/issues/20).It is recommended to pass `default` param to `.dumps` method with something like
```python
def default(obj):
    if isinstance(obj, AnyUrl):
        return str(obj)
    raise TypeError
```

Also `json_encoders` param in pydantic `Config` can't help with this problem, because `AnyUrl` is a subclass of `str` class, so this happens https://github.com/tiangolo/fastapi/blob/master/fastapi/encoders.py#L79, when field is being serialized

I don't see a clear way to fix this. May be we can first try to use `custom_encoder` in `jsonable_encoder` function? Or we need some way to pass `default` to ORJSONResponse, so it could be passed to `.dumps` method.

### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.

Replace each part with your own scenario:

1. Create a file with:

```Python
from fastapi import FastAPI
from pydantic import BaseModel, AnyUrl
from fastapi.responses import ORJSONResponse, UJSONResponse


class SimpleModel(BaseModel):
    value: AnyUrl


app = FastAPI()


@app.get("/json", response_model=SimpleModel)
def read_root():
    return {"value": "http://test"}


@app.get("/ujson", response_model=SimpleModel, response_class=UJSONResponse)
def read_root():
    return {"value": "http://test"}


@app.get("/orjson", response_model=SimpleModel, response_class=ORJSONResponse)
def read_root():
    return {"value": "http://test"}
```

3. Open the browser and call the endpoint `/orjson`.
4. It returns `Internal Server Error`.
5. But I expected it to return `{"value": "http://test"}`.
6. Both `/ujson` and `/json` endpoints work well

### Environment

- OS: macOS
- FastAPI Version: 0.52.0
- Python: 3.7.6

**Comment by requiemofthesouls** [comment]
its also can't encode non string keys in dict :)

**Comment by wkpn** [comment]
> @requiemofthesouls its also can't encode non string keys in dict :)

https://github.com/ijl/orjson#opt_non_str_keys

**Comment by requiemofthesouls** [comment]
> > @requiemofthesouls its also can't encode non string keys in dict :)
> 
> https://github.com/ijl/orjson#opt_non_str_keys

yep

**Comment by connebs** [comment]
As of **orjson** v3.0.0, subclasses of `str`, `int`, `list`, and `dict` are serialized by **orjson**.

This will solve the issue when the dep is bumped in **FastAPI**, yes?

**Comment by raphaelauv** [comment]
since https://github.com/tiangolo/fastapi/commit/3c6dafcc8eac38572028c98b770c800afc1bd1d9

orjson is pinned to >=3.2.1,<4.0.0

So can you close this issue ?

**Comment by fmigneault** [comment]
This still seem broken as of https://github.com/tiangolo/fastapi/releases/tag/0.72.0

I encounter the issue when attempting to convert the OpenAPI specification generated by FastAPI as follows:

```python 
APP = FastAPI(
    contact={
        "url": "https://github.com/tiangolo/fastapi",
    },
)

APP.get("/schema")
async def openapi_schema(request: Request):
    oas = request.app.openapi()
    data = jsonable_encoder(oas)
    text = yaml.safe_dump(data)  # <--- this raises the AnyUrl generated from the contact URL
    resp = Response(text, media_type="text/plain; charset=utf-8")
    return resp
```

**Reply by Miguelme** [reply]
What is the current status ?

**Comment by YuriiMotov** [comment]
This works fine in current version (0.115.14)
![image](https://github.com/user-attachments/assets/16e218d8-a49b-4910-ae6b-c11b5165a382)
